{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9c4ff27",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   position  count\n",
      "0        20      1\n",
      "1        43      1\n"
     ]
    }
   ],
   "source": [
    "from viper.utils import load_mutation_data_from_vcf, get_genome_length_from_fasta\n",
    "from viper.hotspot import sliding_window_mutation_density, detect_hotspots\n",
    "from viper.visualization import plot_mutation_density\n",
    "\n",
    "# Load mutation data and genome length\n",
    "mutation_df = load_mutation_data_from_vcf(\"data/generated_variants.vcf\", min_qual=30)\n",
    "genome_length = get_genome_length_from_fasta(\"data/example_sequences.fasta\")\n",
    "\n",
    "print(mutation_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "906e06e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   start  end  window_midpoint  mutation_count   density\n",
      "0      0   30               15               1  0.033333\n",
      "1      5   35               20               1  0.033333\n",
      "2     10   40               25               1  0.033333\n"
     ]
    }
   ],
   "source": [
    "# Compute mutation density with sliding windows\n",
    "window_size = 30\n",
    "step_size = 5\n",
    "\n",
    "density_df = sliding_window_mutation_density(\n",
    "    mutations_df=mutation_df,\n",
    "    genome_length=genome_length,\n",
    "    window_size=window_size,\n",
    "    step_size=step_size\n",
    ")\n",
    "\n",
    "print(density_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d2ae6e70",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   start  end  window_midpoint  mutation_count   density\n",
      "0      0   30               15               1  0.033333\n",
      "1      5   35               20               1  0.033333\n",
      "2     10   40               25               1  0.033333\n"
     ]
    }
   ],
   "source": [
    "hotspots_df = detect_hotspots(\n",
    "    density_df,\n",
    "    density_threshold=0\n",
    ")\n",
    "\n",
    "print(hotspots_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3ead16f",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_mutation_density(\n",
    "    density_df=density_df,\n",
    "    hotspots_df=hotspots_df,\n",
    "    genome_length=genome_length\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
